crates/hulk_widgets/src/segmented_control.rs

```diff
@@ -0,0 +1,153 @@
+                    state.selected = (state.selected + 1).min(self.selectables.len() - 1);
+                    response.mark_changed();
+                }
+            })
+        }
+        painter.rect_filled(response.rect, rounding, ui.style().visuals.extreme_bg_color);
+
+        let text_rects = text_rects(response.rect, self.selectables.len());
+        let offset = text_rects[0].width();
+
+        let translation = animate_to(self.id, ui.ctx(), offset * state.selected as f32);
+        let selector_rect = text_rects[0].translate(vec2(translation, 0.0)).shrink(2.0);
+        let selector_response =
+            ui.interact(selector_rect, self.id.with("selector"), Sense::click());
+        let selector_style = ui.style().interact(&selector_response);
+        painter.rect_filled(selector_rect, rounding, selector_style.bg_fill);
+
+        let noninteractive_style = ui.style().noninteractive();
+
+        for (idx, (&rect, text)) in text_rects.iter().zip(self.selectables.iter()).enumerate() {
+            let label_response = ui.interact(rect, self.id.with(idx), Sense::click());
+            let style = ui.style().interact(&response);
+
+            let show_line = idx > 0 && state.selected != idx && state.selected + 1 != idx;
+            {
+                let animated_height = ui
+                    .ctx()
+                    .animate_bool(self.id.with("vline").with(idx), show_line);
+
+                let height = vec2(0.0, rect.height() - 4.0);
+                let center = rect.left_center();
+
+                painter.line_segment(
+                    [
+                        center - 0.5 * animated_height * height,
+                        center + 0.5 * animated_height * height,
+                    ],
+                    noninteractive_style.bg_stroke,
+                );
+            }
+
+            if label_response.clicked() {
+                state.selected = idx;
+                response.mark_changed();
+            }
+            painter.text(
+                rect.center(),
+                Align2::CENTER_CENTER,
+                text.to_string(),
+                text_style.clone(),
+                style.text_color(),
+            );
+        }
+        response
+    }
+}
+
+impl<'ui, T: ToString> Widget for SegmentedControl<'ui, T> {
+    fn ui(self, ui: &mut Ui) -> Response {
+        let mut state = load_state(ui.ctx(), self.id);
+        let response = self.show(ui, &mut state);
+        save_state(ui.ctx(), self.id, state);
+        response
+    }
+}
+
+fn load_state(ctx: &Context, id: Id) -> SegmentedControlState {
+    let persisted = ctx.data_mut(|reader| reader.get_temp(id));
+    persisted.unwrap_or_default()
+}
+
+fn save_state(ctx: &Context, id: Id, state: SegmentedControlState) {
+    ctx.data_mut(|writer| writer.insert_temp(id, state));
+}
+
+fn animate_to(source: Id, context: &Context, target: f32) -> f32 {
+    context.animate_value_with_time(source, target, 0.1)
+}
```

Comment:
This is a one-liner function that is used only once. Inline?

Reply:
yes you are completely right, inlined now